packages/cosmic-swingset/lib/ag-solo/vats/lib-wallet.js

```diff
@@ -61,120 +61,96 @@ export async function makeWallet(
+      E(registrar).get(assayId),
+    );
+    const regAssaysP = Promise.all(regAssaysAP);
+    const regUnitsP = Promise.all(
+      payoutRules.map(({ units: { extent } }, i) =>
+        E(regAssaysAP[i]).makeUnits(extent || 0),
+      ),
+    );
+    const purseAssaysP = Promise.all(
+      rawPurses.map(purse => E(purse).getAssay()),
+    );
 
-    const {
-      terms: {
-        assays: [contractAssay0, contractAssay1],
+    const [
+      {
+        terms: { assays: contractAssays },
+        publicAPI,
       },
-      publicAPI,
-    } = await E(zoe).getInstance(instanceHandle);
+      regAssays,
+      regUnits,
+      purseAssays,
+    ] = await Promise.all([instanceP, regAssaysP, regUnitsP, purseAssaysP]);
 
     // =====================
     // === AWAITING TURN ===
     // =====================
 
+    // Order the purses by registered assays.
+    const purses = regAssays.map(regAssay => {
+      const i = purseAssays.indexOf(regAssay);
+      if (i < 0) {
+        return undefined;
+      }
+      return rawPurses[i];
+    });
+
     const invite = await E(publicAPI).makeInvite();
 
     // =====================
     // === AWAITING TURN ===
     // =====================
 
-    assert(contractAssay0 === regAssay0 && regAssay1 === contractAssay1);
-
-    // Check whether we sell on contract assay 0 or 1.
-    const normal = checkOrder(
-      purseAssay0,
-      purseAssay1,
-      contractAssay0,
-      contractAssay1,
+    // Ensure that the offered and contract assays match.
+    contractAssays.forEach((contractAssay, i) =>
+      assert(contractAssay, regAssays[i]),
     );
 
-    const payment0P = E(purse0).withdraw(normal ? purseUnit0 : purseUnit1);
-    const contractUnit0P = E(contractAssay0).makeUnits(extent0 || 0);
-    const contractUnit1P = E(contractAssay1).makeUnits(extent1 || 0);
-
-    const [payment0, contractUnit0, contractUnit1] = await Promise.all([
-      payment0P,
-      contractUnit0P,
-      contractUnit1P,
-    ]);
-
-    // =====================
-    // === AWAITING TURN ===
-    // =====================
-
     // Clone the offer rules to have a writable object.
     const newOfferRules = JSON.parse(JSON.stringify(offerRules));
 
     // Hydrate with the resolved units.
-    newOfferRules.payoutRules[0].units = contractUnit0;
-    newOfferRules.payoutRules[1].units = contractUnit1;
+    newOfferRules.payoutRules.forEach(
+      (payoutRule, i) => (payoutRule.units = regUnits[i]),
+    );
     harden(newOfferRules);
 
-    const payment = normal
-      ? [payment0, undefined, undefined]
-      : [undefined, payment0, undefined];
+    // Look up the payments in the array ordered by registered assays.
+    const payment = await Promise.all(
+      newOfferRules.payoutRules.map(({ kind, units }, i) => {
+        const purse = purses[i];
+        if (kind === 'offerAtMost' && purse) {
+          return E(purse).withdraw(units);
+        }
+        return undefined;
```

Comment:
Is this right? What if the purse is missing for a line with 'offerAtMost'? Wouldn't that be an error?

Reply:
I think it would be up to Zoe to notice it's missing at that point.